arangod/Pregel/Worker/Handler.h

```diff
@@ -350,118 +146,50 @@ struct WorkerHandler : actor::HandlerBase<Runtime, WorkerState<V, E, M>> {
-      } catch (...) {
-        return Result{TRI_ERROR_INTERNAL,
-                      "caught unknown exception when storing graph"};
-      }
-    };
-
-    this->template dispatch<metrics::message::MetricsMessages>(
-        this->state->metricsActor,
-        arangodb::pregel::metrics::message::WorkerStoringFinished{});
-
-    this->template dispatch<conductor::message::ConductorMessages>(
-        this->state->conductor, graphStored());
+    auto newState =
+        this->state->executionState->receive(this->sender, message, dispatcher);
+    changeState(std::move(newState));
 
     return std::move(this->state);
   }
 
-  auto operator()(message::ProduceResults msg)
+  auto operator()(message::ProduceResults message)
       -> std::unique_ptr<WorkerState<V, E, M>> {
-    auto getResults = [this, msg]() -> ResultT<PregelResults> {
-      try {
-        auto storer = std::make_shared<GraphVPackBuilderStorer<V, E>>(
-            msg.withID, this->state->config,
-            this->state->algorithm->inputFormat());
-        storer->store(this->state->magazine).get();
-        return PregelResults{*storer->stealResult()};
-      } catch (std::exception const& ex) {
-        return Result{TRI_ERROR_INTERNAL,
-                      fmt::format("caught exception when receiving results: {}",
-                                  ex.what())};
-      } catch (...) {
-        return Result{TRI_ERROR_INTERNAL,
-                      "caught unknown exception when receiving results"};
-      }
-    };
-    auto results = getResults();
-    this->template dispatch<pregel::message::ResultMessages>(
-        this->state->resultActor,
-        pregel::message::SaveResults{.results = {results}});
-    this->template dispatch<pregel::conductor::message::ConductorMessages>(
-        this->state->conductor,
-        pregel::conductor::message::ResultCreated{.results = {results}});
+    auto newState =
+        this->state->executionState->receive(this->sender, message, dispatcher);
+    changeState(std::move(newState));
 
     return std::move(this->state);
   }
 
-  auto operator()(message::Cleanup msg)
+  auto operator()([[maybe_unused]] message::Cleanup message)
       -> std::unique_ptr<WorkerState<V, E, M>> {
+    this->finish();
+
     LOG_TOPIC("664f5", INFO, Logger::PREGEL)
         << fmt::format("Worker Actor {} is cleaned", this->self);
 
-    this->finish();
-
-    this->template dispatch<pregel::message::SpawnMessages>(
-        this->state->spawnActor, pregel::message::SpawnCleanup{});
-    this->template dispatch<pregel::conductor::message::ConductorMessages>(
-        this->state->conductor, pregel::conductor::message::CleanupFinished{});
-    this->template dispatch<metrics::message::MetricsMessages>(
-        this->state->metricsActor,
-        arangodb::pregel::metrics::message::WorkerFinished{});
+    auto newState =
+        this->state->executionState->receive(this->sender, message, dispatcher);
+    changeState(std::move(newState));
```

Comment:
You have to be able to receive a Cleanup message in every exeution state and cleanup, so you can also think about just reverting the changes in this function and just leave the execution state as it was (or change for better clarity to the Done state I mentioned in ResultsProducedState.cpp) because calling `this->finish()` means that this worker is garbage collected in the near future anyway. Then you would also not need both CleaningUp and CleanedUp execution state.